### ISSUE TITLE
conch.checks.SSHPublicKeyDatabase calls os.seteuid/os.setegid even if it's not necessary

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2607]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)| @therve reported|
|-|-|
|Trac ID|trac#2607|
|Type|enhancement|
|Created|2007-04-26 10:00:15Z|
|Branch|https://github.com/twisted/twisted/tree/conch-non-root-2607-3|

When using manhole, I often only want to access the ssh server with the user running the server (not root). But currently SSHPublicKeyDatabase makes a seteuid everytime, making this impossible by default, whereas I do have access to the authorized_keys file.

Attachments:

* [checkers_2607.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/e9f/e9f2da1cf65392a07f67613715ff9cb23df6ffdf/d344431dce689b2947e14387c9fc89a9c3ef730d.diff) (2292 bytes) - added by therve on 2007-04-26 10:01:03Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__2607 2607
type__enhancement enhancement
reporter__therve therve
priority__high high
milestone__ 
branch__branches_conch_non_root_2607_3 branches/conch-non-root-2607-3
branch_author__ 
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__ 
time__1177581615000000 1177581615000000
changetime__1220339547000000 1220339547000000
version__None None
owner__ 
cc__jml cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_1"></a>|@therve set owner to z3p|
|-|-|

The attached patch corrects the problem; but it would probably need tests...

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve commented|
|-|-|

[#3314](https://github.com/twisted/twisted/issues/3314) has been closed as a duplicate of this one.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve set owner to @therve|
|-|-|

[#3314](https://github.com/twisted/twisted/issues/3314) is not exactly the same because the problem is in twisted.conch.openssh_compat, but it should share code.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve commented|
|-|-|

(In [24051]) Branching to 'conch-non-root-2607'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve commented|
|-|-|

(In [24052]) Tests are hard, see you later.

Refs [#2607](https://github.com/twisted/twisted/issues/2607)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve set owner to @jml|
|-|-|

Ok I think it's not too bad.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @therve|
|-|-|

POSIX has some problems. :(

  * twisted/conch/openssh_compat/factory.py
    * instead of adding `asRoot` as a parameter to `getPrivateKeys`, and instead of doing the seteuid/setegid calls in it, I think the code would be simpler and clearer if you add a helper that calls a function as root and restores the original ids afterwards.  Oops, of course, I guess you still need a flag, to avoid recursing too many times. :/
    * of course, that doesn't fix the problem that this code will call log.msg as root sometimes - hope none of your log files need to be rotated
    * and I hope no other threads resume during the function and do anything that's based on uid/gid.
    * okay those last two probably aren't related to the changes for this ticket
  * twisted/conch/checkers.py
    * if there's a uid/gid helper `getPrivateKeys` uses, `checkKey` could use it as well
  * twisted/conch/test/test_checkers.py
    * nothing seems to pass a value for `MockOS`'s `callback`.  Oh, I see, some code sets the attribute later on.
    * `test_checkKeyAsRoot` should make sure keyFile gets its permissions back to avoid making cleanup of _trial_temp difficult

Thanks for fixing the deprecated keys stuff, too.  Maybe the use of xreadlines could go, too, since xreadlines is basically deprecated and doesn't do much of interest.

Some of the copyright statements seem... torpid?  Can you make the capitalization the same as it is elsewhere? :)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve removed owner|
|-|-|

All points addressed, I think.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set owner to @therve|
|-|-|

  * twisted/python/util.py
    * `runAsEffectiveUser` docstring typos _will_ as _will_
  * twisted/python/test/test_util.py
    * perhaps `MockOS` could be expanded with `setegid` and `seteuid`?  Ideally, I'd like to see these as verified fakes (I _think_ they are correct on Linux, but I'm not sure about other POSIXs).  It'd be nice to at least be able to tell if they are broken on BSD/etc, even if we aren't in a big hurry to make them work.  If we at least keep all this stuff centralized in one place (ie, `MockOS`) then maybe it'll be more likely that someday someone will verify it.
    * The tests don't actually verify that the uid/gid are correct for the duration of the function call, just that they're set and reset.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve removed owner|
|-|-|

I think this is good.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set owner to @therve|
|-|-|

Some tests in `twisted.test.test_process` are failing now. :/

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_12"></a>|@therve removed owner|
|-|-|

Sorry about that, it should be better now.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun set owner to @therve|
|-|-|

  * Can you add module docstrings for:
    * twisted/conch/checkers.py
    * twisted/conch/openssh_compat/factory.py
  * twisted/conch/test/test_checkers.py
    * typo in module docstring
  * twisted/conch/checkers.py
    * some of the indentation in `checkKey` isn't 4 spaces

Some tests need to be skipped on Windows?  Or something:

```
===============================================================================
[ERROR]: twisted.conch.test.test_checkers

Traceback (most recent call last):
  File "c:\twistedbuildbot24\WXP32-full2.4-scmikes-select\Twisted\twisted\trial\runner.py", line 547, in loadPackage
    module = modinfo.load()
  File "c:\twistedbuildbot24\WXP32-full2.4-scmikes-select\Twisted\twisted\python\modules.py", line 380, in load
    return self.pathEntry.pythonPath.moduleLoader(self.name)
  File "c:\twistedbuildbot24\WXP32-full2.4-scmikes-select\Twisted\twisted\python\modules.py", line 618, in moduleLoader
    return self._moduleLoader(modname)
  File "c:\twistedbuildbot24\WXP32-full2.4-scmikes-select\Twisted\twisted\python\reflect.py", line 455, in namedAny
    topLevelPackage = _importAndCheckStack(trialname)
  File "c:\twistedbuildbot24\WXP32-full2.4-scmikes-select\Twisted\twisted\conch\test\test_checkers.py", line 8, in ?
    import os, pwd, base64
exceptions.ImportError: No module named pwd
===============================================================================
[ERROR]: twisted.conch.test.test_openssh_compat

Traceback (most recent call last):
  File "c:\twistedbuildbot24\WXP32-full2.4-scmikes-select\Twisted\twisted\trial\runner.py", line 547, in loadPackage
    module = modinfo.load()
  File "c:\twistedbuildbot24\WXP32-full2.4-scmikes-select\Twisted\twisted\python\modules.py", line 380, in load
    return self.pathEntry.pythonPath.moduleLoader(self.name)
  File "c:\twistedbuildbot24\WXP32-full2.4-scmikes-select\Twisted\twisted\python\modules.py", line 618, in moduleLoader
    return self._moduleLoader(modname)
  File "c:\twistedbuildbot24\WXP32-full2.4-scmikes-select\Twisted\twisted\python\reflect.py", line 455, in namedAny
    topLevelPackage = _importAndCheckStack(trialname)
  File "c:\twistedbuildbot24\WXP32-full2.4-scmikes-select\Twisted\twisted\conch\test\test_openssh_compat.py", line 16, in ?
    from twisted.conch.test.test_checkers import MockOS
  File "c:\twistedbuildbot24\WXP32-full2.4-scmikes-select\Twisted\twisted\conch\test\test_checkers.py", line 8, in ?
    import os, pwd, base64
exceptions.ImportError: No module named pwd
===============================================================================
[ERROR]: twisted.python.test.test_util.RunAsEffectiveUserTests.test_GID

Traceback (most recent call last):
  File "c:\twistedbuildbot24\WXP32-full2.4-scmikes-select\Twisted\twisted\python\test\test_util.py", line 612, in setUp
    self.patch(os, "geteuid", self.mockos.geteuid)
  File "c:\twistedbuildbot24\WXP32-full2.4-scmikes-select\Twisted\twisted\trial\unittest.py", line 928, in patch
    monkeyPatch.patch()
  File "c:\twistedbuildbot24\WXP32-full2.4-scmikes-select\Twisted\twisted\python\monkey.py", line 51, in patch
    self._originals.append((obj, name, getattr(obj, name)))
exceptions.AttributeError: 'module' object has no attribute 'geteuid'
===============================================================================
[ERROR]: twisted.python.test.test_util.RunAsEffectiveUserTests.test_UID

Traceback (most recent call last):
  File "c:\twistedbuildbot24\WXP32-full2.4-scmikes-select\Twisted\twisted\python\test\test_util.py", line 612, in setUp
    self.patch(os, "geteuid", self.mockos.geteuid)
  File "c:\twistedbuildbot24\WXP32-full2.4-scmikes-select\Twisted\twisted\trial\unittest.py", line 928, in patch
    monkeyPatch.patch()
  File "c:\twistedbuildbot24\WXP32-full2.4-scmikes-select\Twisted\twisted\python\monkey.py", line 51, in patch
    self._originals.append((obj, name, getattr(obj, name)))
exceptions.AttributeError: 'module' object has no attribute 'geteuid'
===============================================================================
[ERROR]: twisted.python.test.test_util.RunAsEffectiveUserTests.test_UIDGID

Traceback (most recent call last):
  File "c:\twistedbuildbot24\WXP32-full2.4-scmikes-select\Twisted\twisted\python\test\test_util.py", line 612, in setUp
    self.patch(os, "geteuid", self.mockos.geteuid)
  File "c:\twistedbuildbot24\WXP32-full2.4-scmikes-select\Twisted\twisted\trial\unittest.py", line 928, in patch
    monkeyPatch.patch()
  File "c:\twistedbuildbot24\WXP32-full2.4-scmikes-select\Twisted\twisted\python\monkey.py", line 51, in patch
    self._originals.append((obj, name, getattr(obj, name)))
exceptions.AttributeError: 'module' object has no attribute 'geteuid'
===============================================================================
[ERROR]: twisted.python.test.test_util.RunAsEffectiveUserTests.test_forwardResult

Traceback (most recent call last):
  File "c:\twistedbuildbot24\WXP32-full2.4-scmikes-select\Twisted\twisted\python\test\test_util.py", line 612, in setUp
    self.patch(os, "geteuid", self.mockos.geteuid)
  File "c:\twistedbuildbot24\WXP32-full2.4-scmikes-select\Twisted\twisted\trial\unittest.py", line 928, in patch
    monkeyPatch.patch()
  File "c:\twistedbuildbot24\WXP32-full2.4-scmikes-select\Twisted\twisted\python\monkey.py", line 51, in patch
    self._originals.append((obj, name, getattr(obj, name)))
exceptions.AttributeError: 'module' object has no attribute 'geteuid'
===============================================================================
[ERROR]: twisted.python.test.test_util.RunAsEffectiveUserTests.test_root

Traceback (most recent call last):
  File "c:\twistedbuildbot24\WXP32-full2.4-scmikes-select\Twisted\twisted\python\test\test_util.py", line 612, in setUp
    self.patch(os, "geteuid", self.mockos.geteuid)
  File "c:\twistedbuildbot24\WXP32-full2.4-scmikes-select\Twisted\twisted\trial\unittest.py", line 928, in patch
    monkeyPatch.patch()
  File "c:\twistedbuildbot24\WXP32-full2.4-scmikes-select\Twisted\twisted\python\monkey.py", line 51, in patch
    self._originals.append((obj, name, getattr(obj, name)))
exceptions.AttributeError: 'module' object has no attribute 'geteuid'
===============================================================================
[ERROR]: twisted.python.test.test_util.RunAsEffectiveUserTests.test_takeParameters

Traceback (most recent call last):
  File "c:\twistedbuildbot24\WXP32-full2.4-scmikes-select\Twisted\twisted\python\test\test_util.py", line 612, in setUp
    self.patch(os, "geteuid", self.mockos.geteuid)
  File "c:\twistedbuildbot24\WXP32-full2.4-scmikes-select\Twisted\twisted\trial\unittest.py", line 928, in patch
    monkeyPatch.patch()
  File "c:\twistedbuildbot24\WXP32-full2.4-scmikes-select\Twisted\twisted\python\monkey.py", line 51, in patch
    self._originals.append((obj, name, getattr(obj, name)))
exceptions.AttributeError: 'module' object has no attribute 'geteuid'
===============================================================================
[ERROR]: twisted.python.test.test_util.RunAsEffectiveUserTests.test_takesKeyworkArguments

Traceback (most recent call last):
  File "c:\twistedbuildbot24\WXP32-full2.4-scmikes-select\Twisted\twisted\python\test\test_util.py", line 612, in setUp
    self.patch(os, "geteuid", self.mockos.geteuid)
  File "c:\twistedbuildbot24\WXP32-full2.4-scmikes-select\Twisted\twisted\trial\unittest.py", line 928, in patch
    monkeyPatch.patch()
  File "c:\twistedbuildbot24\WXP32-full2.4-scmikes-select\Twisted\twisted\python\monkey.py", line 51, in patch
    self._originals.append((obj, name, getattr(obj, name)))
exceptions.AttributeError: 'module' object has no attribute 'geteuid'
-------------------------------------------------------------------------------
```

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_14"></a>|@therve removed owner|
|-|-|

Done.

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_15"></a>|@washort set owner to @therve|
|-|-|

twisted/test/test_process.py conflicts with trunk :(

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_16"></a>|@therve commented|
|-|-|

(In [24502]) Branching to 'conch-non-root-2607-2'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_17"></a>|@therve removed owner|
|-|-|

Thanks, solved.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun set owner to @therve|
|-|-|

  * twisted/python/util.py
    * `runAsEffectiveUser` documents `args` and `kwargs` - should it use the `@param *args` and `@param **kwargs` style instead?

Unix is pretty complicated.  I'm not sure this is the right behavior:

```
root@charm:~# python
Python 2.4.3 (#2, Jul 31 2008, 21:56:52)
[GCC 4.0.3 (Ubuntu 4.0.3-1ubuntu5)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import os
>>> from twisted.python.util import runAsEffectiveUser
>>> def report():
...     return (os.getuid(), os.getgid(), os.geteuid(), os.getegid())
...
>>> report()
(0, 0, 0, 0)
>>> os.setegid(1)
>>> os.seteuid(1)
>>> report()
(0, 0, 1, 1)
>>> runAsEffectiveUser(1000, 1000, report)
(0, 0, 1000, 1000)
>>> runAsEffectiveUser(1000, 1, report)
(0, 0, 1000, 1)
>>> runAsEffectiveUser(1, 1000, report)
Traceback (most recent call last):
  File "<stdin>", line 1, in ?
  File "/home/exarkun/Projects/Twisted/trunk/twisted/python/util.py", line 935, in runAsEffectiveUser
    os.setegid(egid)
OSError: [Errno 1] Operation not permitted
>>> report()
(0, 0, 1, 0)
>>>
```

I keep getting confused when I look at that big stack of if/sete{u,g}id calls in `runAsEffectiveUser`.  The tests that include lists of integers are equally or more confusing as well.  I wonder if it would simplify anything to have `MockOS` enforce the same rules as POSIX dictates.  That requires more complexity in `MockOS`, though, and I'm not sure how we could verify that the implementation is correct, since it will require CAP_SETUID which typically isn't available during a test suite run.

I think everything else in the branch looks pretty good now, but I'm a bit worried about the above.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_19"></a>|@therve commented|
|-|-|

(In [24581]) Fix runAsEffectiveUser implementation.

Refs [#2607](https://github.com/twisted/twisted/issues/2607)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_20"></a>|@therve removed owner|
|-|-|

I think the problem is fixed.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_21"></a>|@exarkun set owner to @therve|
|-|-|

If I comment out lines 930, 931, 941, and 942 in `runAsEffectiveUser`:

```
    uid, gid = os.geteuid(), os.getegid()
    if uid == euid and gid == egid:
        return function(*args, **kwargs)
    else:
#        if gid != 0 and gid != egid:
#            os.setegid(0)
        if uid != 0 and (uid != euid or gid != egid):
            os.seteuid(0)
        if egid != 0 and gid != egid:
            os.setegid(egid)
        if euid != 0 and (euid != uid or gid != egid):
            os.seteuid(euid)
        try:
            return function(*args, **kwargs)
        finally:
 #           if egid != 0 and gid != egid:
 #               os.setegid(0)
            if euid != 0 and (uid != euid or gid != egid):
                os.seteuid(0)
            if gid != 0 and gid != egid:
                os.setegid(gid)
            if uid != 0 and (uid != euid or gid != egid):
                os.seteuid(uid)
```

then the function still seems to work (except if I have a non-0 egid and try to run a function with egid 0, the `gid != 0` check below prevents it from being set, but removing that check too seems to make it work completely again).

Most of the unit tests break, though, but I think they're wrong to require the egid to be set to 0 first.  Of course, it's not wrong to set the egid, it's just not necessary.

Everything else still looks fine. ;)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_22"></a>|@therve removed owner|
|-|-|

Security is hard... Thanks for being careful, it should be better now.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_23"></a>|@exarkun set owner to @therve|
|-|-|

Looks good, I think you should merge.

I had difficulty getting a clean test run on Windows.  Twice, there were four errors in `test_process`.  None of the changes to `test_process` in this branch seem like they would have caused such problems.  I was going to try some more test runs but then the Windows slave died.

Build results are here: http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/conch-non-root-2607-2

(You can add num_builds=n to see more results)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_24"></a>|@therve commented|
|-|-|

(In [24647]) Branching to 'conch-non-root-2607-3'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_25"></a>|@therve removed owner|
|-|-|

It seems the error in trunk as well, so it should not block this branch.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_26"></a>|@exarkun set owner to @therve|
|-|-|

Alright, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_27"></a>|@therve set status to `closed`|
|-|-|

(In [24672]) Merge conch-non-root-2607-3

Author: therve
Reviewer: exarkun
Fixes [#2607](https://github.com/twisted/twisted/issues/2607)

Add a new function runAsEffectiveUser, used in conch for now, which allow to
run manhole or the conch server without being root if not necessary.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
